Crash depending on where in a file a certain term is referenced

## Compiler version

3.0.0

## Minification

Unfortunately this is super hard to minimise in isolation but I can provide a minification and reproduction with scalajs-react.

```scala
import japgolly.scalajs.react._

object BUG {

  def problem = Main // crashes

  val NameChanger =
    ScalaComponent.builder[Int]
      .stateless
      .noBackend
      .render_(???)
      .build

  def noProblem = Main // no crash

  val Main =
    ScalaComponent.builder[Unit]
      .initialState(123)
      .render(_ => NameChanger(123))
      .build
}
```

The only difference between `problem` which causes a crash, and `noProblem` which doesn't, is that `problem` is defined before `NameChanger`.

## Reproduction

```sh
git clone https://github.com/japgolly/scalajs-react.git
cd scalajs-react
git checkout tmp/s3crash
sbt compile
```

## Output (click arrow to expand)
<details>

```scala
[info] compiling 1 Scala source to scalajs-react/gh-pages/target/scala-3.0.0/classes ...
[info] exception occurred while typechecking scalajs-react/gh-pages/src/main/scala/ghpages/BUG.scala
[info] exception occurred while compiling scalajs-react/gh-pages/src/main/scala/ghpages/BUG.scala
java.lang.AssertionError: cannot merge Constraint(
 uninstantiated variables: S
 constrained types: 
  [X, P, S]
    (using ev: 
      X => japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[P, S]
    ): 
      Conversion[X, 
        japgolly.scalajs.react.component.builder.ComponentBuilder.Step3[P, S, 
          Unit
        ]
      ]
 bounds: 
     X := 
    japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[Unit, S]
     P := Unit
     S >: (123 : Int)
 ordering: 
) with Constraint(
 uninstantiated variables: 
 constrained types: 
  [X, P, S]
    (using ev: 
      X => japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[P, S]
    ): 
      Conversion[X, 
        japgolly.scalajs.react.component.builder.ComponentBuilder.Step3[P, S, 
          Unit
        ]
      ]
 bounds: 
     X := 
    japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[Unit, Int]
     P := Unit
     S := Int
 ordering: 
), mergeEntries(japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[Unit, S], japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[Unit, Int]) failed while compiling scalajs-react/gh-pages/src/main/scala/ghpages/BUG.scala
[error] ## Exception when compiling 32 sources to scalajs-react/gh-pages/target/scala-3.0.0/classes
[error] java.lang.AssertionError: cannot merge Constraint(
[error]  uninstantiated variables: S
[error]  constrained types: 
[error]   [X, P, S]
[error]     (using ev: 
[error]       X => japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[P, S]
[error]     ): 
[error]       Conversion[X, 
[error]         japgolly.scalajs.react.component.builder.ComponentBuilder.Step3[P, S, 
[error]           Unit
[error]         ]
[error]       ]
[error]  bounds: 
[error]      X := 
[error]     japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[Unit, S]
[error]      P := Unit
[error]      S >: (123 : Int)
[error]  ordering: 
[error] ) with Constraint(
[error]  uninstantiated variables: 
[error]  constrained types: 
[error]   [X, P, S]
[error]     (using ev: 
[error]       X => japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[P, S]
[error]     ): 
[error]       Conversion[X, 
[error]         japgolly.scalajs.react.component.builder.ComponentBuilder.Step3[P, S, 
[error]           Unit
[error]         ]
[error]       ]
[error]  bounds: 
[error]      X := 
[error]     japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[Unit, Int]
[error]      P := Unit
[error]      S := Int
[error]  ordering: 
[error] ), mergeEntries(japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[Unit, S], japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[Unit, Int]) failed
[error] dotty.tools.dotc.core.OrderingConstraint.mergeEntries$1(OrderingConstraint.scala:485)
[error] dotty.tools.dotc.core.OrderingConstraint.$amp$$anonfun$1(OrderingConstraint.scala:508)
[error] dotty.tools.dotc.core.OrderingConstraint.mergeArrays$1$$anonfun$1(OrderingConstraint.scala:460)
[error] scala.runtime.java8.JFunction1$mcVI$sp.apply(JFunction1$mcVI$sp.scala:18)
[error] scala.collection.immutable.Range.foreach(Range.scala:190)
[error] dotty.tools.dotc.core.OrderingConstraint.mergeArrays$2(OrderingConstraint.scala:460)
[error] dotty.tools.dotc.core.OrderingConstraint.merge$2$$anonfun$1(OrderingConstraint.scala:465)
[error] scala.runtime.function.JProcedure2.apply(JProcedure2.java:15)
[error] scala.runtime.function.JProcedure2.apply(JProcedure2.java:10)
[error] dotty.tools.dotc.util.SimpleIdentityMap$Map1.foreachBinding(SimpleIdentityMap.scala:61)
[error] dotty.tools.dotc.core.OrderingConstraint.merge$1(OrderingConstraint.scala:465)
[error] dotty.tools.dotc.core.OrderingConstraint.$amp(OrderingConstraint.scala:508)
[error] dotty.tools.dotc.core.OrderingConstraint.$amp(OrderingConstraint.scala:454)
[error] dotty.tools.dotc.core.TyperState.mergeConstraintWith(TyperState.scala:135)
[error] dotty.tools.dotc.core.TyperState.commit(TyperState.scala:124)
[error] dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:2931)
[error] dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:983)
[error] dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1021)
[error] dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:303)
[error] dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:103)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2669)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2731)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2916)
[error] dotty.tools.dotc.typer.Typer.typeSelectOnTerm$1(Typer.scala:602)
[error] dotty.tools.dotc.typer.Typer.typedSelect(Typer.scala:652)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2639)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2730)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
[error] dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1377)
[error] dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1364)
[error] dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1377)
[error] dotty.tools.dotc.typer.Namer.typedAheadRhs$1$$anonfun$1(Namer.scala:1505)
[error] dotty.tools.dotc.typer.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:225)
[error] dotty.tools.dotc.typer.Namer.typedAheadRhs$2(Namer.scala:1505)
[error] dotty.tools.dotc.typer.Namer.rhsType$1(Namer.scala:1513)
[error] dotty.tools.dotc.typer.Namer.cookedRhsType$1(Namer.scala:1538)
[error] dotty.tools.dotc.typer.Namer.lhsType$1(Namer.scala:1539)
[error] dotty.tools.dotc.typer.Namer.inferredType$1(Namer.scala:1550)
[error] dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1559)
[error] dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:699)
[error] dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:826)
[error] dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:727)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:167)
[error] dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:188)
[error] dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:190)
[error] dotty.tools.dotc.core.Types$TermRef.underlying(Types.scala:2578)
[error] dotty.tools.dotc.core.Types$Type.widen(Types.scala:1201)
[error] dotty.tools.dotc.typer.Typer.simplify(Typer.scala:2744)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2733)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
[error] dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1377)
[error] dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1364)
[error] dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1377)
[error] dotty.tools.dotc.typer.Namer.typedAheadRhs$1$$anonfun$1(Namer.scala:1505)
[error] dotty.tools.dotc.typer.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:225)
[error] dotty.tools.dotc.typer.Namer.typedAheadRhs$2(Namer.scala:1505)
[error] dotty.tools.dotc.typer.Namer.rhsType$1(Namer.scala:1513)
[error] dotty.tools.dotc.typer.Namer.cookedRhsType$1(Namer.scala:1538)
[error] dotty.tools.dotc.typer.Namer.lhsType$1(Namer.scala:1539)
[error] dotty.tools.dotc.typer.Namer.inferredType$1(Namer.scala:1550)
[error] dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1559)
[error] dotty.tools.dotc.typer.Namer.defDefSig(Namer.scala:1642)
[error] dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:703)
[error] dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:826)
[error] dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:727)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:167)
[error] dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:188)
[error] dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:190)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:369)
[error] dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:2611)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2636)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2730)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2822)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2872)
[error] dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2328)
[error] dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2657)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2661)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2730)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2822)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2872)
[error] dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2451)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2702)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2731)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2916)
[error] dotty.tools.dotc.typer.FrontEnd.liftedTree1$1(FrontEnd.scala:79)
[error] dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:84)
[error] dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:43)
[error] dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:85)
[error] dotty.tools.dotc.typer.FrontEnd.runOn$$anonfun$3(FrontEnd.scala:120)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:120)
[error] dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$5(Run.scala:215)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:223)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:230)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:166)
[error] dotty.tools.dotc.Run.compile(Run.scala:150)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:186)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:241)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:176)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:157)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:157)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:204)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:174)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:172)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:457)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:261)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:412)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:499)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:399)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:166)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2362)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2319)
[error] sbt.internal.io.Retry$.apply(Retry.scala:46)
[error] sbt.internal.io.Retry$.apply(Retry.scala:28)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2315)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] java.base/java.lang.Thread.run(Thread.java:834)
[error]            
[error] stack trace is suppressed; run last gh-pages / Compile / compileIncremental for the full output
[error] (gh-pages / Compile / compileIncremental) java.lang.AssertionError: cannot merge Constraint(
[error]  uninstantiated variables: S
[error]  constrained types: 
[error]   [X, P, S]
[error]     (using ev: 
[error]       X => japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[P, S]
[error]     ): 
[error]       Conversion[X, 
[error]         japgolly.scalajs.react.component.builder.ComponentBuilder.Step3[P, S, 
[error]           Unit
[error]         ]
[error]       ]
[error]  bounds: 
[error]      X := 
[error]     japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[Unit, S]
[error]      P := Unit
[error]      S >: (123 : Int)
[error]  ordering: 
[error] ) with Constraint(
[error]  uninstantiated variables: 
[error]  constrained types: 
[error]   [X, P, S]
[error]     (using ev: 
[error]       X => japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[P, S]
[error]     ): 
[error]       Conversion[X, 
[error]         japgolly.scalajs.react.component.builder.ComponentBuilder.Step3[P, S, 
[error]           Unit
[error]         ]
[error]       ]
[error]  bounds: 
[error]      X := 
[error]     japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[Unit, Int]
[error]      P := Unit
[error]      S := Int
[error]  ordering: 
[error] ), mergeEntries(japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[Unit, S], japgolly.scalajs.react.component.builder.ComponentBuilder.Step2[Unit, Int]) failed
```
</details>

 # Workarounds

Any of the following changes avoid the crash:

1.
    ```diff
    -  val NameChanger =
    +  val NameChanger: ScalaComponent[Int, Unit, Unit, CtorType.Props] =
    ```

1.
    ```diff
    -  val Main =
    +  val Main: ScalaComponent[Unit, Int, Unit, CtorType.Nullary] =
    ```

1.
    ```diff
    -      .initialState(123)
    +      .initialState[Int](123)
    ```

1.
    ```diff
           .initialState(123)
    +      .noBackend // this is normally an implicit conversion
    ```

1.
    ```diff
    -      .render(_ => NameChanger(123))
    +      .render(_ => NameChanger.ctor(123)) // this is normally an implicit conversion
    ```

--- a/compiler/src/dotty/tools/dotc/core/Constraint.scala
+++ b/compiler/src/dotty/tools/dotc/core/Constraint.scala
@@ -152,16 +152,6 @@ abstract class Constraint extends Showable {
    */
   def uninstVars: collection.Seq[TypeVar]
 
-  /** The weakest constraint that subsumes both this constraint and `other`.
-   *  The constraints should be _compatible_, meaning that a type lambda
-   *  occurring in both constraints is associated with the same typevars in each.
-   *
-   *  @param otherHasErrors    If true, handle incompatible constraints by
-   *                           returning an approximate constraint, instead of
-   *                           failing with an exception
-   */
-  def & (other: Constraint, otherHasErrors: Boolean)(using Context): Constraint
-
   /** Whether `tl` is present in both `this` and `that` but is associated with
    *  different TypeVars there, meaning that the constraints cannot be merged.
    */

--- a/compiler/src/dotty/tools/dotc/core/Constraint.scala
+++ b/compiler/src/dotty/tools/dotc/core/Constraint.scala
@@ -183,7 +183,4 @@ abstract class Constraint extends Showable {
    *  of athe type lambda that is associated with the typevar itself.
    */
   def checkConsistentVars()(using Context): Unit
-
-  /** A string describing the constraint's contents without a header or trailer */
-  def contentsToString(using Context): String
 }

--- a/compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala
+++ b/compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala
@@ -459,48 +459,6 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
 
 // ----------- Joins -----------------------------------------------------
 
-  def & (other: Constraint, otherHasErrors: Boolean)(using Context): OrderingConstraint = {
-
-    def merge[T](m1: ArrayValuedMap[T], m2: ArrayValuedMap[T], join: (T, T) => T): ArrayValuedMap[T] = {
-      var merged = m1
-      def mergeArrays(xs1: Array[T], xs2: Array[T]) = {
-        val xs = xs1.clone
-        for (i <- xs.indices) xs(i) = join(xs1(i), xs2(i))
-        xs
-      }
-      m2.foreachBinding { (poly, xs2) =>
-        merged = merged.updated(poly,
-            if (m1.contains(poly)) mergeArrays(m1(poly), xs2) else xs2)
-      }
-      merged
-    }
-
-    def mergeParams(ps1: List[TypeParamRef], ps2: List[TypeParamRef]) =
-      ps2.foldLeft(ps1)((ps1, p2) => if (ps1.contains(p2)) ps1 else p2 :: ps1)
-
-    // Must be symmetric
-    def mergeEntries(e1: Type, e2: Type): Type =
-      (e1, e2) match {
-        case _ if e1 eq e2 => e1
-        case (e1: TypeBounds, e2: TypeBounds) => e1 & e2
-        case (e1: TypeBounds, _) if e1 contains e2 => e2
-        case (_, e2: TypeBounds) if e2 contains e1 => e1
-        case (tv1: TypeVar, tv2: TypeVar) if tv1 eq tv2 => e1
-        case _ =>
-          if (otherHasErrors)
-            e1
-          else
-            throw new AssertionError(i"cannot merge $this with $other, mergeEntries($e1, $e2) failed")
-      }
-
-    val that = other.asInstanceOf[OrderingConstraint]
-
-    new OrderingConstraint(
-        merge(this.boundsMap, that.boundsMap, mergeEntries),
-        merge(this.lowerMap, that.lowerMap, mergeParams),
-        merge(this.upperMap, that.upperMap, mergeParams))
-  }.showing(i"constraint merge $this with $other = $result", constr)
-
   def hasConflictingTypeVarsFor(tl: TypeLambda, that: Constraint): Boolean =
     contains(tl) && that.contains(tl) &&
     // Since TypeVars are allocated in bulk for each type lambda, we only have

--- a/compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala
+++ b/compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala
@@ -641,49 +641,10 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
     upperMap.foreachBinding((_, paramss) => paramss.foreach(_.foreach(checkClosedType(_, "upper"))))
   end checkClosed
 
-// ---------- toText -----------------------------------------------------
-
-  private def contentsToText(printer: Printer): Text =
-    //Printer.debugPrintUnique = true
-    def entryText(tp: Type) = tp match {
-      case tp: TypeBounds =>
-        tp.toText(printer)
-      case _ =>
-        " := " ~ tp.toText(printer)
-    }
-    val indent = 3
-    val uninstVarsText = " uninstantiated variables: " ~
-      Text(uninstVars.map(_.toText(printer)), ", ")
-    val constrainedText =
-      " constrained types: " ~ Text(domainLambdas map (_.toText(printer)), ", ")
-    val boundsText =
-      " bounds: " ~ {
-        val assocs =
-          for (param <- domainParams)
-          yield (" " * indent) ~ param.toText(printer) ~ entryText(entry(param))
-        Text(assocs, "\n")
-      }
-    val orderingText =
-      " ordering: " ~ {
-        val deps =
-          for {
-            param <- domainParams
-            ups = minUpper(param)
-            if ups.nonEmpty
-          }
-          yield
-            (" " * indent) ~ param.toText(printer) ~ " <: " ~
-              Text(ups.map(_.toText(printer)), ", ")
-        Text(deps, "\n")
-      }
-    //Printer.debugPrintUnique = false
-    Text.lines(List(uninstVarsText, constrainedText, boundsText, orderingText))
+// ---------- Printing -----------------------------------------------------
 
   override def toText(printer: Printer): Text =
-    Text.lines(List("Constraint(", contentsToText(printer), ")"))
-
-  def contentsToString(using Context): String =
-    contentsToText(ctx.printer).show
+    printer.toText(this)
 
   override def toString: String = {
     def entryText(tp: Type): String = tp match {
@@ -692,7 +653,7 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
     }
     val constrainedText =
       " constrained types = " + domainLambdas.mkString("\n")
-    val boundsText = domainLambdas
+    val boundsText =
       " bounds = " + {
         val assocs =
           for (param <- domainParams)

--- a/compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala
+++ b/compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala
@@ -24,11 +24,13 @@ object OrderingConstraint {
   type ParamOrdering = ArrayValuedMap[List[TypeParamRef]]
 
   /** A new constraint with given maps */
-  private def newConstraint(boundsMap: ParamBounds, lowerMap: ParamOrdering, upperMap: ParamOrdering)(using Context) : OrderingConstraint = {
-    val result = new OrderingConstraint(boundsMap, lowerMap, upperMap)
-    ctx.run.recordConstraintSize(result, result.boundsMap.size)
-    result
-  }
+  private def newConstraint(boundsMap: ParamBounds, lowerMap: ParamOrdering, upperMap: ParamOrdering)(using Context) : OrderingConstraint =
+    if boundsMap.isEmpty && lowerMap.isEmpty && upperMap.isEmpty then
+      empty
+    else
+      val result = new OrderingConstraint(boundsMap, lowerMap, upperMap)
+      ctx.run.recordConstraintSize(result, result.boundsMap.size)
+      result
 
   /** A lens for updating a single entry array in one of the three constraint maps */
   abstract class ConstraintLens[T <: AnyRef: ClassTag] {

--- a/compiler/src/dotty/tools/dotc/core/TyperState.scala
+++ b/compiler/src/dotty/tools/dotc/core/TyperState.scala
@@ -187,11 +187,43 @@ class TyperState() {
    */
   def mergeConstraintWith(that: TyperState)(using Context): Unit =
     that.ensureNotConflicting(constraint)
-    constraint = constraint & (that.constraint, otherHasErrors = that.reporter.errorsReported)
-    for tvar <- constraint.uninstVars do
-      if !isOwnedAnywhere(this, tvar) then includeVar(tvar)
+
+    val comparingCtx =
+      if ctx.typerState == this then ctx
+      else ctx.fresh.setTyperState(this)
+
+    comparing(typeComparer =>
+      val other = that.constraint
+      val res = other.domainLambdas.forall(tl =>
+        // Integrate the type lambdas from `other`
+        constraint.contains(tl) || other.isRemovable(tl) || {
+          val tvars = tl.paramRefs.map(other.typeVarOfParam(_)).collect { case tv: TypeVar => tv }
+          tvars.foreach(tvar => if !tvar.inst.exists && !isOwnedAnywhere(this, tvar) then includeVar(tvar))
+          typeComparer.addToConstraint(tl, tvars)
+        }) &&
+        // Integrate the additional constraints on type variables from `other`
+        constraint.uninstVars.forall(tv =>
+          val p = tv.origin
+          val otherLos = other.lower(p)
+          val otherHis = other.upper(p)
+          val otherEntry = other.entry(p)
+          (  (otherLos eq constraint.lower(p)) || otherLos.forall(_ <:< p)) &&
+          (  (otherHis eq constraint.upper(p)) || otherHis.forall(p <:< _)) &&
+          ((otherEntry eq constraint.entry(p)) || otherEntry.match
+            case NoType =>
+              true
+            case tp: TypeBounds =>
+              tp.contains(tv)
+            case tp =>
+              tv =:= tp
+          )
+        )
+      assert(res || ctx.reporter.errorsReported, i"cannot merge $constraint with $other.")
+    )(using comparingCtx)
+
     for tl <- constraint.domainLambdas do
       if constraint.isRemovable(tl) then constraint = constraint.remove(tl)
+  end mergeConstraintWith
 
   /** Take ownership of `tvar`.
    *

--- a/compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala
+++ b/compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala
@@ -599,6 +599,47 @@ class PlainPrinter(_ctx: Context) extends Printer {
       case _ => "{...}"
     s"import $exprStr.$selectorStr"
 
+  def toText(c: OrderingConstraint): Text =
+    val savedConstraint = ctx.typerState.constraint
+    try
+      // The current TyperState constraint determines how type variables are printed
+      ctx.typerState.constraint = c
+      def entryText(tp: Type) = tp match {
+        case tp: TypeBounds =>
+          toText(tp)
+        case _ =>
+          " := " ~ toText(tp)
+      }
+      val indent = 3
+      val uninstVarsText = " uninstantiated variables: " ~
+        Text(c.uninstVars.map(toText), ", ")
+      val constrainedText =
+        " constrained types: " ~ Text(c.domainLambdas.map(toText), ", ")
+      val boundsText =
+        " bounds: " ~ {
+          val assocs =
+            for (param <- c.domainParams)
+            yield (" " * indent) ~ toText(param) ~ entryText(c.entry(param))
+          Text(assocs, "\n")
+        }
+      val orderingText =
+        " ordering: " ~ {
+          val deps =
+            for {
+              param <- c.domainParams
+              ups = c.minUpper(param)
+              if ups.nonEmpty
+            }
+            yield
+              (" " * indent) ~ toText(param) ~ " <: " ~
+                Text(ups.map(toText), ", ")
+          Text(deps, "\n")
+        }
+      //Printer.debugPrintUnique = false
+      Text.lines(List(uninstVarsText, constrainedText, boundsText, orderingText))
+    finally
+      ctx.typerState.constraint = savedConstraint
+
   def plain: PlainPrinter = this
 
   protected def keywordStr(text: String): String = coloredStr(text, SyntaxHighlighting.KeywordColor)

--- a/compiler/src/dotty/tools/dotc/printing/Printer.scala
+++ b/compiler/src/dotty/tools/dotc/printing/Printer.scala
@@ -154,6 +154,9 @@ abstract class Printer {
   /** Textual representation of info relating to an import clause */
   def toText(result: ImportInfo): Text
 
+  /** Textual representation of a constraint */
+  def toText(c: OrderingConstraint): Text
+
   /** Render element within highest precedence */
   def toTextLocal(elem: Showable): Text =
     atPrec(DotPrec) { elem.toText(this) }

--- a/compiler/src/dotty/tools/dotc/typer/ErrorReporting.scala
+++ b/compiler/src/dotty/tools/dotc/typer/ErrorReporting.scala
@@ -149,7 +149,7 @@ object ErrorReporting {
           "the empty constraint"
         else
           i"""a constraint with:
-             |${c.contentsToString}"""
+             |$c"""
       i"""
         |${TypeComparer.explained(_.isSubType(found, expected), header)}
         |

--- a/compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala
+++ b/compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala
@@ -376,6 +376,7 @@ object ProtoTypes {
     def typedArgs(norm: (untpd.Tree, Int) => untpd.Tree = sameTree)(using Context): List[Tree] =
       if state.typedArgs.size == args.length then state.typedArgs
       else
+        val passedCtx = ctx
         val passedTyperState = ctx.typerState
         inContext(protoCtx.withUncommittedTyperState) {
           val protoTyperState = ctx.typerState
@@ -409,8 +410,7 @@ object ProtoTypes {
                   tvar.instantiate(fromBelow = false)
                 case _ =>
               }
-
-            passedTyperState.mergeConstraintWith(protoTyperState)
+            passedTyperState.mergeConstraintWith(protoTyperState)(using passedCtx)
           end if
           args1
         }

--- a/compiler/src/dotty/tools/dotc/util/SimpleIdentityMap.scala
+++ b/compiler/src/dotty/tools/dotc/util/SimpleIdentityMap.scala
@@ -6,6 +6,7 @@ import collection.mutable.ListBuffer
  *  It has linear complexity for `apply`, `updated`, and `remove`.
  */
 abstract class SimpleIdentityMap[K <: AnyRef, +V >: Null <: AnyRef] extends (K => V) {
+  final def isEmpty: Boolean = this eq SimpleIdentityMap.myEmpty
   def size: Int
   def apply(k: K): V
   def remove(k: K): SimpleIdentityMap[K, V]

--- a/compiler/src/dotty/tools/dotc/util/Stats.scala
+++ b/compiler/src/dotty/tools/dotc/util/Stats.scala
@@ -62,6 +62,7 @@ import collection.mutable
         aggregate()
         println()
         println(hits.toList.sortBy(_._2).map{ case (x, y) => s"$x -> $y" } mkString "\n")
+        hits.clear()
       }
     }
     else op

--- a/compiler/test/dotty/tools/dotc/core/ConstraintsTest.scala
+++ b/compiler/test/dotty/tools/dotc/core/ConstraintsTest.scala
@@ -1,0 +1,54 @@
+package dotty.tools
+package dotc.core
+
+import vulpix.TestConfiguration
+
+import dotty.tools.dotc.core.Contexts.{*, given}
+import dotty.tools.dotc.core.Decorators.{*, given}
+import dotty.tools.dotc.core.Symbols.*
+import dotty.tools.dotc.core.Types.*
+import dotty.tools.dotc.typer.ProtoTypes.constrained
+
+import org.junit.Test
+
+import dotty.tools.DottyTest
+
+class ConstraintsTest:
+
+  @Test def mergeParamsTransitivity: Unit =
+    inCompilerContext(TestConfiguration.basicClasspath,
+        scalaSources = "trait A { def foo[S, T, R]: Any  }") {
+      val tp = constrained(requiredClass("A").typeRef.select("foo".toTermName).info.asInstanceOf[TypeLambda])
+      val List(s, t, r) = tp.paramRefs
+
+      val innerCtx = ctx.fresh.setExploreTyperState()
+      inContext(innerCtx) {
+        s <:< t
+      }
+
+      t <:< r
+
+      ctx.typerState.mergeConstraintWith(innerCtx.typerState)
+      assert(s frozen_<:< r,
+        i"Merging constraints `?S <: ?T` and `?T <: ?R` should result in `?S <:< ?R`: ${ctx.typerState.constraint}")
+    }
+  end mergeParamsTransitivity
+
+  @Test def mergeBoundsTransitivity: Unit =
+    inCompilerContext(TestConfiguration.basicClasspath,
+        scalaSources = "trait A { def foo[S, T]: Any  }") {
+      val tp = constrained(requiredClass("A").typeRef.select("foo".toTermName).info.asInstanceOf[TypeLambda])
+      val List(s, t) = tp.paramRefs
+
+      val innerCtx = ctx.fresh.setExploreTyperState()
+      inContext(innerCtx) {
+        s <:< t
+      }
+
+      defn.IntType <:< s
+
+      ctx.typerState.mergeConstraintWith(innerCtx.typerState)
+      assert(defn.IntType frozen_<:< t,
+        i"Merging constraints `?S <: ?T` and `Int <: ?S` should result in `Int <:< ?T`: ${ctx.typerState.constraint}")
+    }
+  end mergeBoundsTransitivity

--- a/tests/pos/i12730.scala
+++ b/tests/pos/i12730.scala
@@ -1,0 +1,37 @@
+class ComponentSimple
+
+class Props {
+  def apply(props: Any): Any = ???
+}
+
+class Foo[C] {
+  def build: ComponentSimple = ???
+}
+
+class Bar[E] {
+  def render(r: E => Any): Unit = {}
+}
+
+trait Conv[A, B] {
+  def apply(a: A): B
+}
+
+object Test {
+  def toComponentCtor[F](c: ComponentSimple): Props = ???
+
+  def defaultToNoBackend[G, H](ev: G => Foo[H]): Conv[Foo[H], Bar[H]] = ???
+
+  def conforms[A]: A => A = ???
+
+  def problem = Main // crashes
+
+  def foo[H]: Foo[H] = ???
+
+  val NameChanger =
+    foo
+      .build
+
+  val Main =
+    defaultToNoBackend(conforms).apply(foo)
+      .render(_ => toComponentCtor(NameChanger)(13))
+}

